### PR TITLE
Bug fix sending messages

### LINKED ISSUE #1: 1) Bug fix: Sending messages
There are some bugs in our starting code that we would like you to investigate and resolve. First, when you send a new message, it does not immediately appear on the screen. We would like new messages to be immediately added to the chat UI for both existing conversations and new conversations. (Note: This may require more than one change)
Additionally, messages are not displayed in the correct order on page load. Messages should be displayed in order with the oldest messages at the top and newest messages at the bottom.

The starting code has some Cypress tests for this ticket (bug-fix-ticket.spec.js). Please ensure that your code passes these tests before requesting review on your PR for this feature (the README has details on how to run the tests).

### PR DESCRIPTION
**Pull request by benyakirten**
New messages weren't correctly updating in the client for two reasons.
1. postMessage in thunkCreators contacted the backend to create a new message. However, it did not let the promise resolve and still used the data.
2. addMessageToStore and addNewConvoToStore mutated the redux state, preventing the components from updating.

Messages weren't sorted correctly because when a conversation loaded, they were retrieved from the database in descending order. However, the client side should display the messages in ascending order.

### CODE REVIEW

File: client/src/store/utils/reducerFunctions.js [hunk 1]
```diff
@@ -13,9 +13,14 @@ export const addMessageToStore = (state, payload) => {
...
-      convo.latestMessageText = message.text;
-      return convo;
+      return {
+        ...convo,
+        latestMessageText: message.text,
+        messages: [
+          ...convo.messages,
+          message
+        ]
+      }
```
Comment (andreluciani) [inline client/src/store/utils/reducerFunctions.js]: :100:

File: client/src/store/utils/reducerFunctions.js [hunk 2]
```diff
@@ -69,10 +74,15 @@ export const addSearchedUsersToStore = (state, users) => {
...
-      return convo;
+      return {
+        ...convo,
+        id: message.conversationId,
+        latestMessageText: message.text,
+        messages: [
+          ...convo.messages,
+          message
+        ]
+      }
```
Comment (andreluciani) [inline client/src/store/utils/reducerFunctions.js:88]: :100:

File: client/src/store/utils/thunkCreators.js [hunk 3]
```diff
@@ -93,9 +93,9 @@ const sendMessage = (data, body) => {
 
 // message format to send: {recipientId, text, conversationId}
 // conversationId will be set to null if its a brand new conversation
-export const postMessage = (body) => (dispatch) => {
+export const postMessage = (body) => async (dispatch) => {
   try {
-    const data = saveMessage(body);
+    const data = await saveMessage(body);
```
Comment (andreluciani) [inline client/src/store/utils/thunkCreators.js:98]: :100:

File: server/routes/api/conversations.js [hunk 4]
```diff
@@ -19,9 +19,9 @@ router.get("/", async (req, res, next) => {
         },
       },
       attributes: ["id"],
-      order: [[Message, "createdAt", "DESC"]],
+      order: [[Message, "createdAt", "ASC"]],
       include: [
-        { model: Message, order: ["createdAt", "DESC"] },
+        { model: Message, order: ["createdAt", "ASC"] },
```
Comment (andreluciani) [inline server/routes/api/conversations.js]: Good! But this alone, even though solves the messages order, breaks two other things:

- The latest text message on the sidebar is wrong
- The order of the conversations on the sidebar is wrong too

**Review by andreluciani — COMMENTED** [review]
Nice job!
Take a look at the comment I've left below, address it and let us know on Slack when the PR is ready for another review!

File: client/src/store/utils/reducerFunctions.js [hunk 5]
```diff
@@ -11,15 +11,23 @@ export const addMessageToStore = (state, payload) => {
...
-    if (convo.id === message.conversationId) {
-      convo.messages.push(message);
-      convo.latestMessageText = message.text;
-      return convo;
-    } else {
-      return convo;
-    }
-  });
+  // the conversation with the latest message should go to the top of the conversations
+  // sorting (the simpler solution) is O(n log(n)) but this solution is O(n)
```
Comment (FilipeGorodscy) [inline client/src/store/utils/reducerFunctions.js:15]: No need to worry about dynamically updating the conversations on the Sidebar! You can remove this comment here :)

File: server/routes/api/conversations.js [hunk 6]
```diff
@@ -69,6 +69,9 @@ router.get("/", async (req, res, next) => {
 
       // set properties for notification count and latest message preview
       convoJSON.latestMessageText = convoJSON.messages[0].text;
+
+      // we need the conversations in DESC order but need the messages in ASC order
+      convoJSON.messages.reverse();
```
Comment (FilipeGorodscy) [inline server/routes/api/conversations.js:74]: 💯 You can remove the comment :)

**Review by FilipeGorodscy — APPROVED** [review]
Awesome work! 👍

**Pull request merged**